### PR TITLE
Solver participation guard rework

### PR DESCRIPTION
**Pull request by squadgazzz**
# Description
During the previous tests, I found a [major issue](https://github.com/cowprotocol/infrastructure/pull/2748#issuecomment-2698560775) with the db-statistics-based strategy. In short, it was based on a SQL query that had a sliding window nature of the results, which, in some cases, requires much more time for solvers to recover, which is unacceptable. This PR changes the whole approach and switches to the in-memory-based stats.

# Changes

- [ ] At a fresh restart, an SQL query fetches the last N competitions metadata to populate the in-memory cache effectively. Already at this point, some solvers might be banned.
- [ ] Then, the guard's stats get updated on each settlement result.
- [ ] The solver ban period is no longer defined in time but in the settlements count(10 by default). Basically, once the solver is banned, it skips N next settlements.
- [ ] If there is the only one remaining active solver, avoid banning it to keep the competition going.
- [ ] A very important detail is the failure rate management 👇 
Consider the following example:
The failure rate after which the solver gets banned is 30%. The current stats: 2 failures and 2 total attempts, making the failure rate to be 100%. On the next success, the failure rate becomes 66%, which is still above the threshold. That makes the solver to get banned again. This is avoided by not increasing the failure rate once it is reached. When recovered, the rate is adjusted to become below the threshold and as much as close to it, so on the subsequent failure, the solver gets banned again.

## How to test
It was very painful to make the e2e tests stable, but currently, it works. A driver mock had to be implemented to achieve that.
Also, I couldn't avoid adding unit tests, since some logic requires edge case checking, which is impossible to achieve with e2e tests. At first, the feature will be deployed in the log/metrics-mode to observe statistics.

### CODE REVIEW

File: crates/autopilot/src/arguments.rs [hunk 1]
```diff
@@ -253,26 +253,15 @@ pub struct Arguments {
...
 #[derive(Debug, clap::Parser)]
 pub struct DbBasedSolverParticipationGuardConfig {
-    /// Enables or disables the solver participation guard
-    #[clap(
-        id = "db_enabled",
-        long = "db-based-solver-participation-guard-enabled",
-        env = "DB_BASED_SOLVER_PARTICIPATION_GUARD_ENABLED",
-        default_value = "true"
-    )]
-    pub enabled: bool,
```
Comment (squadgazzz) [inline crates/autopilot/src/arguments.rs:263]: This switch was useless initially since each guard has its own switch.

File: crates/autopilot/src/arguments.rs [hunk 2]
```diff
@@ -282,11 +271,14 @@ pub struct NonSettlingSolversFinderConfig {
         id = "non_settling_solvers_blacklisting_enabled",
         long = "non-settling-solvers-blacklisting-enabled",
         env = "NON_SETTLING_SOLVERS_BLACKLISTING_ENABLED",
-        default_value = "true"
+        default_value = "true",
+        action = ArgAction::Set, // allow to provide the value in the command args
+        value_parser = clap::value_parser!(bool) // required for the above
```
Comment (squadgazzz) [inline crates/autopilot/src/arguments.rs:276]: The value is enabled by default, so to disable it, an explicit value needs to be supported.

File: crates/autopilot/src/domain/competition/participation_guard/statistics.rs [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+
+        if pending == 0 {
+            // The metric is updated regardless the config is enabled to track the
+            // statistics.
+            Metrics::get()
+                .banned_solver
+                .with_label_values(&[driver.name.as_ref(), ban_reason.as_str()])
+                .inc();
+
+            if !at_least_two_remaining {
```
Comment (m-lord-renkse) [inline crates/autopilot/src/domain/competition/participation_guard/statistics.rs:324]: should we have a minimum of running solvers configurable?

**Review by m-lord-renkse — COMMENTED** [review]
I did a relatively quick review, I need to digest it and do another round later on.

File: crates/autopilot/src/domain/competition/participation_guard/statistics.rs [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+/// settlements.
+#[derive(Clone)]
+pub(super) struct SolverValidator(Arc<Inner>);
+
+struct Inner {
+    persistence: infra::Persistence,
+    /// A map of banned solver addresses and the corresponding number of
+    /// settlements still need to skip before the solver can participate in
+    /// the competition again.
+    banned_solvers: dashmap::DashMap<eth::Address, u32>,
```
[the remark this answers is not in the document]
Comment (squadgazzz) [reply crates/autopilot/src/domain/competition/participation_guard/statistics.rs:26]: Yes

File: crates/autopilot/src/domain/competition/participation_guard/statistics.rs [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+            }
+        };
+        let mut competitions_tracker = self.0.competitions_tracker.lock().await;
+        metadata.into_iter().for_each(|m| {
+            competitions_tracker.update(m);
+        });
+
+        let low_settling_solvers = competitions_tracker.find_high_failure_solvers();
+        let consecutive_failed_solvers = competitions_tracker.find_consecutive_failed_solvers();
+
```
[the remark this answers is not in the document]
Comment (squadgazzz) [reply crates/autopilot/src/domain/competition/participation_guard/statistics.rs:272]: Technically, yes, but I am not sure why we should care since this happens in a _blocking_ init function. Also, the next operations are not async.

File: crates/autopilot/src/domain/competition/participation_guard/statistics.rs [hunk 6]
```diff
@@ -0,0 +1,503 @@
...
+            while let Some(metadata) = competition_updates_receiver.recv().await {
+                tracing::debug!(?metadata, "competition metadata received");
+                // Decrease the pending settlements counter for all banned solvers
+                for mut entry in self_.0.banned_solvers.iter_mut() {
+                    *entry.value_mut() = entry.value().saturating_sub(1);
+                }
+                let solver = metadata.solver;
+                let mut lock = self_.0.competitions_tracker.lock().await;
+                lock.update(metadata);
+
```
[the remark this answers is not in the document]
Comment (squadgazzz) [reply crates/autopilot/src/domain/competition/participation_guard/statistics.rs:239]: Why do we need it? This code works in a loop...

File: crates/autopilot/src/domain/competition/participation_guard/statistics.rs (hunk 3 above)
Comment (squadgazzz) [reply crates/autopilot/src/domain/competition/participation_guard/statistics.rs]: Makes sense. Added a config.

File: crates/autopilot/src/domain/competition/participation_guard/statistics.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (squadgazzz) [reply crates/autopilot/src/domain/competition/participation_guard/statistics.rs:246]: Actually, the mutex now looks useless. It can probably be dropped.

File: crates/autopilot/src/domain/competition/participation_guard/statistics.rs [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+        if success {
+            self.total += 1;
+            // Recover fast in case the solver is banned.
+            if self.failed > 0 {
+                // Keep the failure rate at the threshold level, so the solver is
+                // able to participate again. And at the same time, on the next failure it gets
+                // banned again.
+                while self.failure_rate() >= high_failure_threshold {
+                    self.total += 1;
+                }
```
Comment (MartinquaXD) [inline crates/autopilot/src/domain/competition/participation_guard/statistics.rs:55]: This seems super sketchy. The higher the `total` the less individual successes or failures change the ratio which seems problematic.
If the idea is to implement a sliding window I would expect `total` to be constant so `failed` would have to be reduced instead. Although I'm not sure this is even overall the best solution. Have to think more about that.

File: crates/autopilot/src/domain/competition/participation_guard/statistics.rs [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+            }
+        } else if !success && self.failure_rate() <= high_failure_threshold {
+            // Keep the failure rate at the same level, so the solver is
+            // able to recover fast with a single successful settlement.
+            self.failed += 1;
+            self.total += 1;
+        }
```
Comment (MartinquaXD) [inline crates/autopilot/src/domain/competition/participation_guard/statistics.rs:62]: Is it intended that `total` will not get increased in the `else` case? The naming `total` seems weird when there are cases when it doesn't get increased and cases where it gets increased multiple times (see above comment).

File: crates/autopilot/src/domain/competition/participation_guard/statistics.rs [hunk 9]
```diff
@@ -0,0 +1,511 @@
...
+    /// The maximum size of the queue where the oldest metadata will be removed.
+    max_cache_size: usize,
+    /// The threshold to consider the solver as a low-settling solver. Once the
+    /// threshold is reached, it stops growing on consecutive failed settlements
+    /// to make is possible for solver to recover with a single successful
+    /// settlement.
+    high_failure_threshold: f64,
+    /// The minimum number of competitions the solver should win to be
+    /// considered as low-settling.
+    min_won_competitions: u32,
```
Comment (MartinquaXD) [inline crates/autopilot/src/domain/competition/participation_guard/statistics.rs:97]: The doc comment reads a bit weird. I think it's supposed to express that we need at least `n` data points before the mechanism is allowed to kick in but it reads like winning competitions makes them bad solvers that need to get punished.

File: crates/autopilot/src/domain/competition/participation_guard/statistics.rs [hunk 10]
```diff
@@ -0,0 +1,511 @@
...
+    }
+}
+
+/// Keeps track of the last N competitions metadata to find solvers with high
+/// failure rates and consecutive failed settlements.
+/// Keeping only the last N competitions metadata is considered more efficient
+/// for solvers to recover from a temporary failure state.
+struct CompetitionsTracker {
+    /// A FIFO queue to store the last N competitions metadata.
+    queue: VecDeque<competition::Metadata>,
```
Comment (MartinquaXD) [inline crates/autopilot/src/domain/competition/participation_guard/statistics.rs:87]: Why do we need to store the metadata at all? It seems like we can keep our per solver metrics up-to-date without it, no?

File: crates/autopilot/src/arguments.rs [hunk 11]
```diff
@@ -248,31 +248,24 @@ pub struct Arguments {
...
-    /// cache.
-    #[clap(long, env, default_value = "5m", value_parser = humantime::parse_duration)]
-    pub solver_blacklist_cache_ttl: Duration,
+    /// The minimum number of active solvers to stop the banning mechanism.
+    #[clap(long, env, default_value = "1")]
+    pub min_active_solvers_threshold: u32,
 
     #[clap(flatten)]
-    pub non_settling_solvers_finder_config: NonSettlingSolversFinderConfig,
+    pub non_settling_solvers_finder: NonSettlingSolversFinderConfig,
```
Comment (MartinquaXD) [inline crates/autopilot/src/arguments.rs:265]: I still think detecting solvers that fail repeatedly after another is a failure case that we don't need to detect.
Removing this altogether could make things simpler overall.

File: crates/autopilot/src/domain/competition/participation_guard/statistics.rs [hunk 12]
```diff
@@ -0,0 +1,511 @@
...
+            dto::notify::BanReason::UnsettledConsecutiveAuctions => {
+                self.0.config.non_settling_solvers_finder.enabled
+            }
+        };
+        let total_solvers_count = self.0.drivers_by_address.len() as u32;
+        let banned_solvers_count = self
+            .0
+            .banned_solvers
+            .iter()
+            .fold(0, |acc, entry| acc + (*entry.value() > 0) as u32);
```
Comment (MartinquaXD) [inline crates/autopilot/src/domain/competition/participation_guard/statistics.rs:318]: Seems like this can be easier expressed with a `.filter()` and `.count()`.

File: crates/autopilot/src/arguments.rs (hunk 11 above)
Comment (squadgazzz) [reply crates/autopilot/src/arguments.rs:265]: The [original issue](https://github.com/cowprotocol/services/issues/3221) highlights the case when a well-settling solver can suddenly start providing solutions that can't be settled for some reason. For example, if in the last 100 auctions the settlement rate was, say, 80% of success, it will take a lot of time for the solver to reach the high-error settling threshold, which is 90%. Without the non-settling guard, we won't be able to block the solver right away.

File: crates/autopilot/src/domain/competition/participation_guard/statistics.rs (hunk 10 above)
Comment (squadgazzz) [reply crates/autopilot/src/domain/competition/participation_guard/statistics.rs:87]: The statistics get updated once an item leaves the queue. This is required to accumulate the stat for the last 100 events only. In case we continue only accumulating, the stat can grow a lot and it will take a very long time to detect the low settling solvers.

File: crates/autopilot/src/domain/competition/participation_guard/statistics.rs (hunk 8 above)
Comment (squadgazzz) [reply crates/autopilot/src/domain/competition/participation_guard/statistics.rs:62]: It is intended. The else path means the settlement failed and the failure rate is already above the threshold. In that case we don't change the failure rate to be able to recover fast on the first successful settlement. I'll think how to express this behavior better since the comment 2 lines above states the same but doesn't work for some reason.

File: crates/autopilot/src/arguments.rs (hunk 2 above)
Comment (squadgazzz) [reply crates/autopilot/src/arguments.rs:280]: Without this, providing the value doesn't work. By default, it tracks the persistence of the `non-settling-solvers-blacklisting-enabled` string without any value. When the string is not provided and the default value is `true`, there is no way to disable it without the explicit value parser.

File: crates/autopilot/src/domain/competition/participation_guard/statistics.rs [hunk 13]
```diff
@@ -0,0 +1,511 @@
...
+    persistence: infra::Persistence,
+    /// A map of banned solver addresses and the corresponding number of
+    /// settlements still need to skip before the solver can participate in
+    /// the competition again.
+    banned_solvers: dashmap::DashMap<eth::Address, u32>,
+    /// Stop banning solvers when the number of remaining active
+    /// solvers hits this threshold. This is crucial to prevent the protocol
+    /// from blocking itself by banning all the solvers since the recover
+    /// mechanism requires new competitions.
+    min_active_solvers_threshold: u32,
```
[the remark this answers is not in the document]
Comment (squadgazzz) [reply crates/autopilot/src/domain/competition/participation_guard/statistics.rs:31]: Currently, the cache gets updated only once a settlement result is received. For the proposed approach, a cache entry needs to be evicted on each empty auction, right?

File: crates/database/src/solver_competition.rs [hunk 14]
```diff
@@ -109,109 +109,38 @@ GROUP BY sc.id
...
-WHERE total_wins >= $3 AND (1 - (total_settlements::decimal / NULLIF(total_wins, 0))) > $4;
-    "#;
+WITH last_auctions AS (
+    SELECT ca.id as auction_id, ps.solver
+    FROM competition_auctions ca
+    JOIN proposed_solutions ps ON ps.auction_id = ca.id
+    WHERE ca.deadline < $1 AND ps.is_winner = true
+    ORDER BY ca.id DESC
+    LIMIT $2
+)
```
[the remark this answers is not in the document]
Comment (squadgazzz) [reply crates/database/src/solver_competition.rs:132]: Doesn't it support multiple winners? All of them will be marked as `is_winner=true`, right? So it should work. At least, when implementing the feature, I tried to keep multiple winners in mind. Also, some tests should cover that.

[the remark this answers is not in the document]
Comment (sunce86) [reply crates/database/src/solver_competition.rs:132]: I meant if multiple winners feature is ON, and you want last 10 auctions, this functions won't return last 10 auctions but last 10 winners, meaning less than 10 auctions in theory. So not actually a big issue, just a comment to be aware of.

[the remark this answers is not in the document]
Comment (squadgazzz) [reply crates/database/src/solver_competition.rs:132]: True, I need to rename the function, agree.

File: crates/autopilot/src/run_loop.rs [hunk 15]
```diff
@@ -335,8 +335,20 @@ impl RunLoop {
...
-                .await
+                .await;
+
+            if let Err(err) = self_
+                .competition_updates_sender
+                .send(competition::Metadata {
+                    auction_id,
+                    solver,
+                    settled: result.is_ok(),
+                })
```
[the remark this answers is not in the document]
Comment (squadgazzz) [reply crates/autopilot/src/run_loop.rs:346]: The logic tracks results for each solver individually, so I can't think of a better place. Especially when each settlement is executed separately.

File: crates/autopilot/src/arguments.rs (hunk 11 above)
Comment (MartinquaXD) [reply crates/autopilot/src/arguments.rs:265]: This seems mostly like an issue of finding good parameters, no?
Let's say we configure a sliding window of 20 submissions and a problematic error rate of 50%. In the absolutely worst case a solver can only halt the protocol for 2 minutes.

To me it seems like the least complex solution would be to configure the error threshold strategy in a way that it would trigger similarly quickly as the consecutive errors strategy.

**Review by MartinquaXD — COMMENTED** [review]
This PR currently quite big. Could you split this up a bit more. Also I'm still not sure we need to have the additional complexity of counting consecutive failures. AFAIU we can approximate that error detection with sufficiently small window sizes in the general failure rate detection.

Also there is some complexity coming from the driver proxy error helper logic. Do you think this can be reduced with unit tests?
